src/Java.Interop.Tools.JavaTypeSystem/JavaTypeCollection.cs

```diff
@@ -0,0 +1,259 @@
+
+		// Expose ReadOnly versions so internal type management cannot be bypassed
+		public IReadOnlyDictionary<string, JavaPackage> Packages => packages;
+		public IReadOnlyDictionary<string, JavaTypeModel> Types => types;
+
+		public IReadOnlyDictionary<string, JavaTypeModel> TypesFlattened => types_flattened;
+
+		// We only keep a flattened version of reference types. The main issue is that the Managed nesting
+		// may not match the Java nesting (ie: types nested in Java interfaces that C# originally didn't support).
+		// Since we don't actually *need* this model to be nested it's simpler to keep them flattened.
+		public IReadOnlyDictionary<string, JavaTypeModel> ReferenceTypesFlattened => reference_types_flattened;
+
+		public string? ApiSource { get; set; }
+		public string? Platform { get; set; }
+
+		public JavaTypeCollection ()
+		{
+			built_in_types.Add ("void", new JavaBuiltInType ("void"));
+			built_in_types.Add ("boolean", new JavaBuiltInType ("boolean"));
+			built_in_types.Add ("int", new JavaBuiltInType ("int"));
+			built_in_types.Add ("byte", new JavaBuiltInType ("byte"));
+			built_in_types.Add ("double", new JavaBuiltInType ("double"));
+			built_in_types.Add ("float", new JavaBuiltInType ("float"));
+			built_in_types.Add ("long", new JavaBuiltInType ("long"));
+			built_in_types.Add ("short", new JavaBuiltInType ("short"));
+			built_in_types.Add ("char", new JavaBuiltInType ("char"));
+		}
+
+		/// <summary>
+		/// Adds a new package with the specified name.  Note if package already exists, existing package will be returned.
+		/// </summary>
+		public JavaPackage AddPackage (string name, string jniName, string? managedName = null)
+		{
+			if (packages.TryGetValue (name, out var pkg))
+				return pkg;
+
+			var new_pkg = new JavaPackage (name, jniName, managedName);
+
+			packages.Add (new_pkg.Name, new_pkg);
+
+			return new_pkg;
+		}
+
+		/// <summary>
+		/// Adds a type to the collection.  Note parent classes must be added before nested classes.
+		/// </summary>
+		/// <returns>True if type was added to collection. False if type could not be added because its parent type was missing.</returns>
+		public bool AddType (JavaTypeModel type)
+		{
+			var nested_name = type.NestedName;
+
+			// Not a nested type
+			if (!nested_name.Contains ('.')) {
+				types [type.FullName] = type;
+
+				types_flattened [type.FullName] = type;
+
+				return true;
+			}
+
+			var full_name = type.FullName.ChompLast ('.');
+
+			// Nested type, find parent model to put it in
+			if (types_flattened.TryGetValue (full_name, out var parent)) {
+				if (!parent.NestedTypes.Contains (type))
+					parent.NestedTypes.Add (type);
+
+				type.ParentType = parent;
+				types_flattened [type.FullName] = type;
+
+				return true;
+			}
+
+			// Could not find parent type to nest child type in
+			return false;
+		}
+
+		/// <summary>
+		/// Adds a reference type to the collection.
```

Comment:
This nomenclature is somewhat confusing, in that there is already a concept of a ["reference type"](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/reference-types).

Perhaps this should be named `AddReferencedType()` (note addition of `d`).

Reply:
Nice, I like that.

Reply:
oh dear, I suggested "referenced type", and now no longer remember what it means!